src/api/providers/kilocode.ts

```diff
@@ -20,19 +20,62 @@ export class KiloCodeHandler extends BaseProvider implements SingleCompletionHan
 		})
 	}
 
-	async *createMessage(systemPrompt: string, messages: Anthropic.Messages.MessageParam[]): ApiStream {
+	private getIdempotencyKey(taskId: string, checkpointNumber: number): string {
+		// Create a deterministic idempotency key based on task_id and checkpoint number
+		return `${taskId}-${checkpointNumber}`
+	}
+
+	async *createMessage(
+		systemPrompt: string,
+		messages: Anthropic.Messages.MessageParam[],
+		taskId?: string,
+		checkpointNumber?: number,
+	): ApiStream {
 		let stream: AnthropicStream<Anthropic.Messages.RawMessageStreamEvent>
 		const cacheControl: CacheControlEphemeral = { type: "ephemeral" }
-		let { id: modelId, maxTokens, thinking, temperature, virtualId } = this.getModel()
+		const { id: modelId, maxTokens, thinking, temperature, virtualId } = this.getModel()
 
-		const userMsgIndices = messages.reduce(
-			(acc, msg, index) => (msg.role === "user" ? [...acc, index] : acc),
-			[] as number[],
-		)
+		// Use a for loop instead of reduce with spread to avoid linting error
+		const userMsgIndices: number[] = []
+		for (let i = 0; i < messages.length; i++) {
+			if (messages[i].role === "user") {
+				userMsgIndices.push(i)
+			}
+		}
 
 		const lastUserMsgIndex = userMsgIndices[userMsgIndices.length - 1] ?? -1
 		const secondLastMsgUserIndex = userMsgIndices[userMsgIndices.length - 2] ?? -1
 
+		// Prepare request options with headers
+		const requestOptions: { headers: Record<string, string> } = (() => {
+			const betas: string[] = []
+
+			// Check for models that support prompt caching
+			switch (modelId) {
+				case "claude-3-7-sonnet-20250219":
+				case "claude-3-5-sonnet-20241022":
+				case "claude-3-5-haiku-20241022":
+				case "claude-3-opus-20240229":
+				case "claude-3-haiku-20240307":
+					betas.push("prompt-caching-2024-07-31")
+					break
+			}
+
+			const headers: Record<string, string> = {}
+
+			// Add beta features if any
+			if (betas.length > 0) {
+				headers["anthropic-beta"] = betas.join(",")
+			}
+
+			// Add idempotency key if task_id and checkpoint number are provided
+			if (taskId && checkpointNumber !== undefined) {
+				headers["idempotency-key"] = this.getIdempotencyKey(taskId, checkpointNumber)
+			}
```

Comment:
This is the AI way of doing it, the more efficient (preventing multiple object mutations) and readable way (imho) to do it would be like this:
```typescript
const headers = {
  'anthropic-beta': betas.length > 0 ? betas.join(",") : undefined,
  'idempotency-key': taskId && checkpointNumber !== undefined ? this.getIdempotencyKey(taskId, checkpointNumber),
}
```